quantum/sequencer/sequencer.c

```diff
@@ -0,0 +1,299 @@
+#include "sequencer.h"
+
+# if defined(SEQUENCER_ENABLE)
+
+#ifdef MIDI_ENABLE
+#include "process_midi.h"
+#endif
+
+sequencer_config_t sequencer_config = {
+    false,      // enabled
+    { false },  // steps
+    { 0 },      // track notes
+    60,         // tempo
+    SQ_RES_4,   // resolution
+};
+
+uint8_t sequencer_active_tracks = 0;
+uint8_t sequencer_current_track = 0;
+uint8_t sequencer_current_step = 0;
+uint16_t sequencer_timer = 0;
+
+/**
+ * Because Digital Audio Workstations get overwhelmed when too many MIDI signals are sent concurrently,
+ * We use a "phase" state machine to delay some of the events.
+ */
+enum sequencer_phase_t {
+    SEQUENCER_PHASE_ATTACK,     // t=0ms, send the MIDI note on signal
+    SEQUENCER_PHASE_RELEASE,    // t=SEQUENCER_PHASE_RELEASE_TIMEOUT ms, send the MIDI note off signal
+    SEQUENCER_PHASE_PAUSE       // t=step duration ms, loop
+};
+
+uint8_t sequencer_phase = SEQUENCER_PHASE_ATTACK;
+
+bool is_sequencer_on(void) {
+    return sequencer_config.enabled;
+}
+
+void sequencer_on(void) {
+    dprintln("sequencer on");
```

Comment:
```suggestion
#ifdef CONSOLE_ENABLE
    dprintln("sequencer on");
#endif
```

Reply:
Shouldn't `dprintln()` et al be compiled out if console is not enabled?

Reply:
Weird, I've had instances where it's complained. Ignore these, then.